Rudimentary undo mechanism on clean_squashed_page_numbers

mewlware opened the pull request:
Right now undo mechanism is only implemented on clean_squashed_page_numbers

I've looked at all the other cleaning functions, it seems like I can implement undo in the exact same way. Would love it if you can give early feedback to this one implementation first to influence the rest!

mewlware committed:
Rudimentary undo mechanism on clean_squashed_page_numbers

JohnAmadeo commented on indolaw-parser/parser_ui.py:
```diff
@@ -24,8 +24,9 @@ def print_dashed_line():
-def print_yes_no():
-    print(f"{colored('y', 'green')} / {colored('n', 'red')}")
+def print_yes_no_undo():
+    print(f"Press {colored('z', 'blue')} to undo")
```
can we make 'u' undo? the 'z' choice is pretty clever (since cmd+z right?) but 'u' is closer to 'y' and 'n' on the keyboard and less awkward to type (since z is usually the responsibility of the left pinky)

JohnAmadeo replied:
also if we change it to 'u' we can get rid of this line, and in the line below we can just make it '(u)ndo'

JohnAmadeo commented on indolaw-parser/parser_utils.py:
```diff
@@ -847,29 +847,48 @@ def clean_squashed_page_numbers(law: List[str]) -> List[str]:
+    last_idx = len(law)
+    last_input_idx = 0
+
+    while idx < last_idx:
```
i think it's clearer if you just do `while idx < len(law)` since you only ever use the `last_idx` variable once anyways

mewlware replied:
If I use `while idx < len(law)` wouldn't that mean the python recalculates `len(law)` for every loop?

JohnAmadeo replied:
good point; pulling it out into a variable is fine then

(altho, technically len(law) != last_idx semantically, since the length of an array is technically 1 larger than the last index of the array due to 0-based indexing)

JohnAmadeo commented on indolaw-parser/parser_utils.py:
```diff
@@ -847,29 +847,48 @@ def clean_squashed_page_numbers(law: List[str]) -> List[str]:
             elif user_input == 'n':
                 new_law.append(line)
+            elif user_input == 'z':
+                for delete_idx in range(idx - 1, last_input_idx - 1, -1):
+                    new_law.pop(delete_idx)
```
qq: instead of passing an argument to `new_law.pop` can't we just call `pop()` w/o arguments since we basically want to remove all lines from most recent until `last_input_idx`

![](https://media2.giphy.com/media/WxIXZWVuZFRmlgwu8b/giphy.gif)

mewlware replied:
😲 that's true! Okay will do!

JohnAmadeo reviewed (commented):
Approach for "simple cases" looks good Mel!

For the more complicated cases (copy-pasting your own words from the chat below)

> i don't know which is the lesser evil here. take example where the function that takes the user input returns an int value. should i:
> 1. return -1 (seems unstable because i know -1 can mean a lot of things in python)
> 2. return an entirely different type (i've had very small experience in Go and C, I'd prefer it if the return value has consistent types)
> OR
> 3. change the return value to a tuple/dict? instead of just returning an int it would be a (str, int) with (undo, int) or (no_undo, int) or something similar

- I agree none of these options are "beautiful", but I think you listed out all the ones that makes sense
- Of the 3, I like returning a tuple of (enum, whatever is needed) the most. In my mind, it can look stg like this (as an e.g)

```
def f():
   # code...
   if nothing wrong:
       return UserInputType.NORMAL_ANSWER, 45
   elif something wrong:
       return UserInputType.UNDO_OVERRIDE, None

def function_that_calls_f():
   user_input_type, user_input = f()
   if user_input_type == UserInputType.UNDO_OVERRIDE:
       # handle override
   elif user_input_type == UserInputType.NORMAL_ANSWER:
      # proceed as usual
   else:
      raise Exception('something something')

```

mewlware committed:
Adds undo mechanism to a few more functions + infinite undo capability

Adds infinite undo mechanism to:

1. insert_perubahan_section_open_quotes
2. def insert_perubahan_section_close_quotes(law: List[str]) -> List[str]:
3. insert_penjelasan_perubahan_section_open_quotes
4. insert_penjelasan_perubahan_section_close_quotes
5. clean_split_plaintext

Extends rudimentary undo to infinite undo to:
1. clean_squashed_page_numbers

mewlware committed:
Bug fixes and general cleanup

Bug fixes on insert_perubahan_section_open_quotes and insert_penjelasan_perubahan_section_open_quotes

JohnAmadeo commented on indolaw-parser/parser_utils.py:
```diff
@@ -618,14 +623,26 @@ def insert_perubahan_section_open_quotes(law: List[str]) -> List[str]:
+                    print_max_undo()
+                #deletes the OPEN_QUOTE_CHAR if the user input in the index is 'y'
+                elif i_input[-1]['input'] == 'y':
+                    new_law[i_input[-1]['i']] = new_law[i_input[-1]['i']][1:]
+                    i = i_input[-1]['i']
```
1) Instead of calling `i_input[-1]['i']` 3x, can we assign it to a separate variable before and give it a nice descriptive name? :)

JohnAmadeo commented on indolaw-parser/parser_utils.py:
```diff
@@ -600,8 +600,13 @@ def insert_perubahan_section_open_quotes(law: List[str]) -> List[str]:
+    i = 0
+    last_i = len(law)
+    # i_input contains a dicts of input index and the associated input
+    i_input = []
```
could we name this something more explicit, maybe `user_input_history`? :) (the comment is good though 👍 )

JohnAmadeo replied:
also instead of having the field be named `i`, would something like `line_index` be clearer? (this is my fault hehe for using `i` in the loop in the first place, but since it's outside the context of a loop feels even more important to be clear)

JohnAmadeo replied:
also `user_input` instead of `input`?

JohnAmadeo commented on indolaw-parser/parser_utils.py:
```diff
@@ -644,7 +661,13 @@ def insert_perubahan_section_close_quotes(law: List[str]) -> List[str]:
+    i = 0
+    last_i = len(open_quote_indexes)
+    # i_input contains the index where CLOSE_QUOTE_CHAR is added
+    i_input = []
```
same comment as above

JohnAmadeo commented on indolaw-parser/parser_utils.py:
```diff
@@ -719,7 +753,16 @@ def insert_penjelasan_perubahan_section_open_quotes(law: List[str]) -> List[str]
+    i = 0
+    #i_input contains dicts of index of line where input is needed
+    #and its associated input
+    i_input = []
```
same comment as above

JohnAmadeo commented on indolaw-parser/parser_utils.py:
```diff
@@ -767,7 +822,12 @@ def insert_penjelasan_perubahan_section_close_quotes(law: List[str]) -> List[str
-    for i, _ in enumerate(open_quote_indexes):
+    i = 0
+    # i_input contains a list of index where close quotes is added
+    i_input = []
```
same comment as above

JohnAmadeo commented on indolaw-parser/parser_utils.py:
```diff
@@ -797,8 +857,17 @@ def insert_penjelasan_perubahan_section_close_quotes(law: List[str]) -> List[str
             if best_guess_index == -1:
                 raise Exception(
-                    'Invalid input - not best guess available to use')
+                    'Invalid input - no best guess available to use')
```
![](https://media0.giphy.com/media/l49JThpgIOD24fGxi/giphy.gif)

good catch!

JohnAmadeo commented on indolaw-parser/parser_utils.py:
```diff
@@ -847,29 +919,49 @@ def clean_squashed_page_numbers(law: List[str]) -> List[str]:
+                else:
+                    for _ in range(idx - 1, idx_input[-1] - 1, -1):
+                        new_law.pop()
+                    print(f"{idx_input}")
```
what is this printed for?

JohnAmadeo commented on indolaw-parser/parser_utils.py:
```diff
@@ -847,29 +919,49 @@ def clean_squashed_page_numbers(law: List[str]) -> List[str]:
+    
+    idx = 0
+    last_idx = len(law)
+    idx_input = []
```
same comment as above

JohnAmadeo commented on indolaw-parser/parser_utils.py:
```diff
@@ -880,21 +972,29 @@ def clean_split_plaintext(law: List[str]) -> List[str]:
-        previous_line_long = i > 0 and len(law[i-1]) > 20
-        previous_line_not_all_caps = i > 0 and not law[i-1].isupper()
-        not_all_caps = not law[i].isupper()
+        line = law[idx]
+
+        really_long = len(law[idx]) > 75
+        long_enough = len(law[idx]) > 5
+        starts_with_lowercase = law[idx][0].islower()
+        starts_with_number = law[idx][0].isnumeric()
+        previous_line_long = idx > 0 and len(law[idx-1]) > 20
+        previous_line_not_all_caps = idx > 0 and not law[idx-1].isupper()
+        not_all_caps = not law[idx].isupper()
```
since you already have `line` can you replace all uses of `law[idx]` below with `line`?

JohnAmadeo commented on indolaw-parser/parser_utils.py:
```diff
@@ -880,21 +972,29 @@ def clean_split_plaintext(law: List[str]) -> List[str]:
+    idx = 0
+    last_idx = len(law)
+    # idx_input contains dicts of input index and the associated input
+    idx_input = []
```
same comment as above

JohnAmadeo commented on indolaw-parser/parser_utils.py:
```diff
@@ -905,36 +1005,58 @@ def clean_split_plaintext(law: List[str]) -> List[str]:
+                    for idx_del in range(idx - 1, idx_input[-1]['idx'] - 1, -1):
+                        if idx_del == idx_input[-1]['idx'] and idx_input[-1]['input'] == 'y':
+                            continue
+                        print(f"deleting line on idx {idx_del}")
```
Do we need this?

JohnAmadeo reviewed (approved):
Logic looks good Mel! Have some small nits re: variable naming, but that's basically it & feel free to merge afterwards
![](https://media1.giphy.com/media/kBZBlLVlfECvOQAVno/giphy.gif)
